geospaas_processing/downloaders.py

```diff
@@ -71,94 +150,84 @@ def extract_file_name_from_response(response):
-            file_name = url.split('/')[-1]
-        file_name = '_'.join([name for name in [file_prefix, file_name] if name])
-        if not file_name:
-            raise ValueError(
-                "No file name could be extracted from the request and no file prefix was provided")
-        return file_name
+    def download_file(cls, file, url, connection):
+        """Download the file using the Response object contained in the
+        `connection` argument
+        """
+        chunk = None
+        for chunk in connection.iter_content(chunk_size=cls.CHUNK_SIZE):
+            file.write(chunk)
+        else:
+            # This executes after the loop and raises an error if the
+            # response is unexpectedly empty like it sometimes happens
+            # with scihub
+            if chunk is None:
+                raise DownloadError(f"Getting an empty file from '{url}'")
+
+
+class FTPDownloader(Downloader):
+    """Downloader for FTP repositories"""
 
     @classmethod
-    def check_and_download_url(cls, url, download_dir, file_prefix='', **kwargs):
-        """
-        Download a file from a HTTP URL if it doesn't already exist.
-        The return value is a two-elements tuple containing:
-        - the name of the file
-        - a boolean which is True if the file was downloaded, and False if it was already present
+    def connect(cls, url, auth=(None, None)):
+        """Connects to the remote FTP repository.
+        Returns a ftplib.FTP object.
         """
-        auth = cls.build_basic_auth(kwargs)
         try:
-            response = requests.get(url, stream=True, auth=auth)
-            response.raise_for_status()
-        # Raising DownloadError enables to display a clear message in the API response
-        except requests.RequestException as error:
+            return ftplib.FTP(host=urlparse(url).netloc, user=auth[0], passwd=auth[1])
+        except ftplib.all_errors as error:
             raise DownloadError(f"Could not download from '{url}'") from error
 
-        # Sometimes scihub's response is empty
-        if len(response.content) == 0:
-            raise DownloadError(f"Getting an empty file from '{url}'")
-
-        # early quit if file already exists
-        file_name = cls.get_file_name(response, url, file_prefix)
-        file_path = os.path.join(download_dir, file_name)
-        if os.path.exists(file_path) and os.path.isfile(file_path):
-            return file_name, False
-
-        # Try to free some space if we can get the size of the file about to be downloaded
-        file_size = cls.get_remote_file_size(response, auth)
-        if file_size:
-            LOGGER.debug("Checking there is enough free space to download %s bytes", file_size)
-            utils.LocalStorage(path=download_dir).free_space(file_size)
+    @classmethod
+    def get_file_name(cls, url, connection):
+        """Extracts the file name from the URL"""
+        return urlparse(url).path.split('/')[-1] or None
```

Comment:
use os.path.basename and dirname if needed instead of`split('/')[-1]`

https://docs.python.org/3/library/os.path.html#os.path.dirname

Reply:
I am dealing with a URL, not a path, so this is not OS dependent.
By using `os.path` I would have problems on OS which do not use '/' as a path separator.